Render patterns in volume segments to increase distinguishability

philippotto opened the pull request:
- adds patterns to volume segments with varying angles, pattern shapes (stripes vs. grid), frequencies
- changes the color generation code to use a more predictable approach
- uses the jet color scheme for the segments

### URL of deployed dev instance (used for testing):
- https://segmentationpatterns.webknossos.xyz

### Steps to test:
- open a volume tracing or dataset with segmentations and inspect the patterns
- play with the "pattern opacity" setting in the segmentation layer settings
- zoom in and out and ensure that the frequency of the stripes changes to pleasant values when the mag is switched
- check that the color in the "segmentation" tab matches the rendered color

### Issues:
- fixes #4240 

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Needs datastore update after deployment
- [X] Ready for review

philippotto commented:
@daniel-wer I think, you can delay the review until we decide that the feature itself is really ready for merging. The feedback until now wasn't that enthusiastic.

philippotto commented on frontend/javascripts/oxalis/view/right-menu/mapping_info_view.js:
```diff
@@ -66,17 +66,30 @@ type State = {
-  const lastEightBits = id & (2 ** 8 - 1);
-  const computedColor = (lastEightBits * goldenRatio) % 1.0;
-  const value = customColors != null ? customColors[lastEightBits] || 0 : computedColor;
+  const last8Bits = id % 2 ** 8;
+  const last16Bits = id % 2 ** 16;
+
+  const colorCount = 17;
+  const colorSeed = 1.41421;
+  const colorIndex = getElementOfPermutation(last16Bits, colorCount, colorSeed);
+  const colorValue = (1.0 / colorCount) * colorIndex;
+
+  const value = customColors != null ? customColors[last8Bits] || 0 : colorValue;
```
This part still needs to be adapted to the primitive root approach.

philippotto commented:
@daniel-wer Ok, I think this is ready-for-review-ish. I still need to adapt the table in the "segmentation tab" so that the new color generation is used there, too. But I think the rest is ready to review and test.

Thanks to @valentin-pinkau's suggestion, I also revised the pseudo-random generation of pattern features to use primitive roots which gives us an elegant approach without collisions within these sequences. I added a detailed explanation to the source code and hope this makes sense :)

daniel-wer commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.js:
```diff
@@ -7,29 +7,149 @@ import { hsvToRgb } from "./utils.glsl";
+    }
+
     vec3 convertCellIdToRGB(vec4 id) {
       float golden_ratio = 0.618033988749895;
```
It looks like this is no longer used

daniel-wer commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.js:
```diff
@@ -7,29 +7,149 @@ import { hsvToRgb } from "./utils.glsl";
+      // Also, when zooming out, coordinates change faster which make the pattern more turbulent. Dividing by the
+      // zoomValue compensates this. Note that the zoomStep should not be used, because that value relates to the
+      // three-dimensional dataset. Since the patterns are only 2D, the zoomValue is a better proxy.
+      float finedTunedScale = 0.15;
```
Should this be `fineTunedScale`?

daniel-wer commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.js:
```diff
@@ -7,29 +7,149 @@ import { hsvToRgb } from "./utils.glsl";
+      float sequenceValue = mod(pow(primitiveRoot, oneBasedIndex), sequenceLength);
+
+      return
+        // Only use sequenceLength if request element is not the first of the sequence
```
```suggestion
        // Only use sequenceLength if request element is the first of the sequence
```

daniel-wer replied:
However, I feel like this would be easier to understand without the not in `isNotFirstElement`.
Does `float isFirstElement = float(oneBasedIndex == 1.0);` work? (from https://stackoverflow.com/questions/13674637/glsl-cast-bool-to-float)

daniel-wer commented:
One more thing: The isosurface color seems to still be the old color as well. I don't think the isosurfaces need the patterns, but the color should be equal :)

philippotto commented:
> I reviewed most of the code and added some questions and suggestions. I didn't review the part involving the scaling of the world coordinates in detail (lines 107+ in segmentation.glsl) as my brain shut off trying. I'm not sure whether it's needed, though as it works very well +1

I added some explaining comments there and hope it's easier to grasp now :)

> I would propose to add a general description of what kinds of patterns, angles and colors we're using now - a little high-level description of what's going on.

Good idea! Added that now.

> I'm not sure whether it's only me, but I find it very hard to distinguish most of the patterns with the default pattern opacity of 40 (in datasets with a full segmentation layer). A value like 70 works a lot better for me. Not a big deal, as it's configurable, but I wanted to throw this suggestion in here :)

Yeah, I noticed that the ideal segmentation pattern opacity value highly depends on the actual segmentation opacity value (so, 40 works well for me when the volume opacity is higher). This is not really ideal, but I couldn't come up with a quick solution which improves this. In the end, most users should fine tune this, anyway, I think.

> One more thing: The isosurface color seems to still be the old color as well. I don't think the isosurfaces need the patterns, but the color should be equal :)

Yeah, that was the last missing piece (also affected the segmentation tab table). It was a bit tricky to get right, since the color calculation suffers from imprecisions. In the end, I ensured that the JS implementation suffers from the same imprecisions :see_no_evil: Since it's consistent, it seems fine to me. I also changed the color count so that the used primitive root is 2 instead of 3, since this yields better precision.

@MichaelBuessemeyer Since @daniel-wer is on vacation, could you give this PR a final pass (test it and also look over the code)? Thank you!

philippotto commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.js:
```diff
@@ -7,29 +7,167 @@ import { hsvToRgb } from "./utils.glsl";
+      float useGrid = step(mod(getElementOfPermutation(significantSegmentIndex, useGridSequenceLength, 2.0), 2.0), 0.5);
+      float aaStripeValueA = aa_step(stripeValueA);
+      float aaStripeValueB = aa_step(stripeValueB);
+      float aaStripeValue = 1.0 - max(aaStripeValueA, useGrid * aaStripeValueB);
```
The two stripes A and B are orthogonal to each other. I `useGrid` is true, these orthogonal stripes are combined (by taking the max of both), so that a grid is produced. If `useGrid` is false, only the first stripe A is used (so, there is a simple stripe pattern).
I'm not sure how to make the naming more clear, but I'll add an explanation to the two stripe value definitions so that this is easier to understand.

MichaelBuessemeyer replied:
> The two stripes A and B are orthogonal to each other. I useGrid is true, these orthogonal stripes are combined (by taking the max of both), so that a grid is produced. If useGrid is false, only the first stripe A is used (so, there is a simple stripe pattern).

Wow thats really smart :man_student:

philippotto commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.js:
```diff
@@ -7,29 +7,167 @@ import { hsvToRgb } from "./utils.glsl";
+      // to ones and zeros. This has the benefit that the periodicity has a prime length.
+      float useGridSequenceLength = 13.;
+      float useGrid = step(mod(getElementOfPermutation(significantSegmentIndex, useGridSequenceLength, 2.0), 2.0), 0.5);
+      float aaStripeValueA = aa_step(stripeValueA);
+      float aaStripeValueB = aa_step(stripeValueB);
+      float aaStripeValue = 1.0 - max(aaStripeValueA, useGrid * aaStripeValueB);
+
+      vec4 HSV = vec4(
+        colorValue,
+        1.0 - 0.5 * ((1. - aaStripeValue) * segmentationPatternOpacity / 100.0),
+        1.0 - 0.5 * (aaStripeValue * segmentationPatternOpacity / 100.0),
+        1.0
+      );
```
Hmm, I'm not too convinced of taking "gray" into the semantics, since the stripes are basically created by building an accompanying color to the primary color. If the pattern opacity is 100%, that secondary color is gray, but if the opacity is below 100% it's not really gray. That's why, I'd stick to `stripeValue`, but I'll add a comment which explains what this means.

MichaelBuessemeyer replied:
Maybe just adding a comment that explains, that the grid is constructed by using the combination of two stripe patters should be enough

philippotto commented on frontend/javascripts/oxalis/view/right-menu/mapping_info_view.js:
```diff
@@ -66,19 +66,78 @@ type State = {
+  const oneBasedIndex = (index % sequenceLength) + 1.0;
+  const isFirstElement = oneBasedIndex === 1.0;
+
+  // The GLSL implementation of pow is 2**(y * log2(x)) in which
+  // intermediate results can suffer from precision loss. The following
+  // code mimics this behavior to get a consistent coloring in GLSL and
+  // JS.
+  const imprecise = x => new Float32Array([x])[0];
+  function glslPow(x, y) {
+    return Math.floor(imprecise(2 ** (y * imprecise(Math.log2(x)))));
+  }
+  const sequenceValue = glslPow(primitiveRoot, oneBasedIndex) % sequenceLength;
+
+  // Only use sequenceLength if the requested element is the first of the sequence
+  // Otherwise, return the actual sequenceValue
+  return isFirstElement ? sequenceLength : sequenceValue;
+}
+
+// Input in [0,1]
+// Output in [0,1] for r, g and b
+function colormapJet(x: number): [number, number, number] {
+  const r = _.clamp(x < 0.89 ? (x - 0.35) / 0.31 : 1.0 - ((x - 0.89) / 0.11) * 0.5, 0, 1);
+  const g = _.clamp(x < 0.64 ? (x - 0.125) * 4.0 : 1.0 - (x - 0.64) / 0.27, 0, 1);
+  const b = _.clamp(x < 0.34 ? 0.5 + (x * 0.5) / 0.11 : 1.0 - (x - 0.34) / 0.31, 0, 1);
+
+  return [r, g, b];
+}
+
+// From: https://stackoverflow.com/a/54070620/896760
+// Input: r,g,b in [0,1], out: h in [0,360) and s,v in [0,1]
+function rgb2hsv(rgb: [number, number, number]): [number, number, number] {
+  const [r, g, b] = rgb;
+  const v = Math.max(r, g, b);
+  const n = v - Math.min(r, g, b);
+
+  // eslint-disable-next-line no-nested-ternary
+  const h = n !== 0 && (v === r ? (g - b) / n : v === g ? 2 + (b - r) / n : 4 + (r - g) / n);
+  // $FlowIgnore
+  return [60 * (h < 0 ? h + 6 : h), v && n / v, v];
+}
```
Very good points! I was a bit sloppy there :see_no_evil: I moved the code into the utils.js module. For the two color-related functions, I also chose `utils.js`, since the `color_generator.js` file is not glsl-related (the glsl code uses the `.glsl.js` file extension). I could have created a new file for these two functions, but I think they are also alright in the utils module :)

MichaelBuessemeyer replied:
I think you mistook this file with the shader file. I meant to move the none-shader implementation (`mapping_info_view.js`) of the three helper functions to other files.


Nevertheless, also moving the shader implementation is / was a good idea :smile:

philippotto replied:
Ah, I see :) Yeah, I got that bit wrong. I moved the JS code now too, however, I just decided to move them directly to the glsl counter-parts. Since these two implementations should always stay in sync, I thought, it would make sense to keep them in the same place. Do you agree? Since their existence is mostly GLSL-motivated, I think, it's fine to put them to the glsl spot.

MichaelBuessemeyer replied:
Sounds reasonable :+1:

philippotto commented:
@MichaelBuessemeyer Thanks for your feedback! I just incorporated it. Could you have another look?

MichaelBuessemeyer commented:
Alright, the last thing that might need to be done is updating the screenshots :framed_picture:. After that let's merge this awesomeness :man_dancing:
